src/main/takeDetection/takeDetection.ts

```diff
@@ -48,56 +48,202 @@ export function findSentences(words: Word[]): Sentence[] {
+  if (isSimilarTake)
+    updatedCurrentSentenceIdx =
+      potentialTakeLen * potentialTakeStartIdxs.length;
+
+  return { isSimilarTake, updatedCurrentSentenceIdx };
+};
+
+const newTakeGroup = (
+  potentialTakeStartIdxs: number[],
+  potentialTakeLen: number,
+  sentences: Sentence[]
+): InjectableTakeGroup => {
+  // if all the sentence in potential take are similar, add them to takeGroups
   const newGroup: InjectableTakeGroup = { takes: [] };
-  let nextTakeRange: IndexRange = {
-    startIndex: sentenceOne.startIndex as number,
-    endIndex: sentenceOne.endIndex as number,
-  };
-  let nextTake: InjectableTake = { wordRange: nextTakeRange };
-  newGroup.takes.push(nextTake);
-  nextTakeRange = {
-    startIndex: sentenceTwo.startIndex as number,
-    endIndex: sentenceTwo.endIndex as number,
-  };
-  nextTake = { wordRange: nextTakeRange };
-  newGroup.takes.push(nextTake);
+
+  potentialTakeStartIdxs.forEach((startIdx) => {
+    const takeStartSentence = sentences[startIdx];
+    const takeEndSentence = sentences[startIdx + potentialTakeLen - 1];
+
+    const takeRange: IndexRange = {
+      startIndex: takeStartSentence.startIndex as number,
+      endIndex: takeEndSentence.endIndex as number,
+    };
+    const take: InjectableTake = { wordRange: takeRange };
+    newGroup.takes.push(take);
+  });
+
   return newGroup;
-}
+};
 
-export function findTakes(words: Word[], threshold = THRESHOLD): InjectableTakeGroup[] {
-  const takeGroups: InjectableTakeGroup[] = [];
-  const sentences: Sentence[] = findSentences(words);
-  for (let i = 0; i < sentences.length - 1; i += 1) {
-    const similarity: number = getSimilarityScore(
-      sentences[i].sentenceString,
-      sentences[i + 1].sentenceString
+const startDetection: (
+  potentialTakeLen: number,
+  potentialTakeStartIdxs: number[],
+  sentences: Sentence[],
+  takeGroups: InjectableTakeGroup[]
+) => number = (
+  potentialTakeLen,
+  potentialTakeStartIdxs,
+  sentences,
+  takeGroups
+) => {
+  // check similarity between sentences in each take chunk
+  const { isSimilarTake, updatedCurrentSentenceIdx } =
+    detectPotentialSimilarTakes(
+      potentialTakeLen,
+      potentialTakeStartIdxs,
+      sentences
     );
-    if (similarity > threshold) {
-      if (takeGroups.length !== 0) {
-        const lastTake: IndexRange = takeGroups.at(-1)?.takes.at(-1)
-          ?.wordRange as IndexRange;
-        if (
-          lastTake.startIndex === sentences[i].startIndex &&
-          lastTake.endIndex === sentences[i].endIndex
-        ) {
-          const nextTakeRange: IndexRange = {
-            startIndex: sentences[i + 1].startIndex as number,
-            endIndex: sentences[i + 1].endIndex as number,
-          };
-          const nextTake: InjectableTake = { wordRange: nextTakeRange };
-          takeGroups.at(-1)?.takes.push(nextTake);
-        } else {
-          takeGroups.push(newTakeGroup(sentences[i], sentences[i + 1]));
+
+  if (isSimilarTake)
+    takeGroups.push(
```

Comment:
In general mutating the input to a function makes it hard to reason about what functions do - it's much easier if they are pure i.e. they take defined inputs and give defined outputs, without modifying the inputs they are given. Can you think of a way to apply that to this function?

(I'm aware the push method is asymptotically faster but I think in this case readability > speed as take detection should be pretty fast anyway)